### PR TITLE
test: Add smoke tests and CI jobs for testing

### PR DESCRIPTION
**Pull request by locnnil**
- Add test documentation
- Add CI jobs to configure snap specific tasks
- Add tests

### CODE REVIEW

File: test/rt-cyclictest [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+. common
+
+name=$(basename "$0")
+program="${name#rt-}"
+test_init $program
+
+printc "WHITE" "Starting...\n"
+
+output=$(cyclictest --mlockall --smp --priority=80 --interval=200 --distance=0 -q -l 10 2>/dev/null)
```
Comment (farshidtz) [inline test/rt-cyclictest:13]: Use long form, for better readability.

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -1,8 +1,8 @@
-name: Snap Builder
+name: Snap build and test
 
 on:
   push:
-    branches: [ main ]
+    branches: [ main, IENG-1072-smoke-tests ]
```
Comment (farshidtz) [inline .github/workflows/build.yaml:5]: To be reverted before merging.

File: test/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# RT-tests smoke tests 
+
+These tests are smoke tests that check if the mainly functionalities of each program is working.
```
Comment (farshidtz) [inline test/README.md:3]:
```suggestion
These are smoke tests that check if the main functionalities of each program is working.
```

File: test/run [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+for test in "${tests[@]}"; do
+	echo "$test ($[${#passed[@]} + ${#failed[@]} + 1]/${#tests[@]})"
+	./$test && passed=(${passed[@]} $test) || failed=(${failed[@]} $test)
+	echo
+done
+
+echo "SUMMARY:"
+echo "  TOTAL  $[${#passed[@]} + ${#failed[@]}]"
+echo "  PASSED ${#passed[@]}"
+echo "  FAILED ${#failed[@]}  (${failed[@]})"
```
Comment (farshidtz) [inline test/run:24]: This kind of accounting makes more sense if one single application is being tested and if we are interested in knowing the total or ratio of passing/failing tests.

I'd change to summarize:
- PASSED: list of applications
- FAILED: list of applications
- SKIPPED: list of applications

File: test/rt-cyclicdeadline [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+# Check program output behavior
+final_table='T:\s*\d+\s*\(\s*\d+\s*\)\s*I:\d+\s*C:\s*\d+\s*Min:\s*\d+\s*Act:\s*\d+\s*Avg:\s*\d+\s*Max:\s*\d+'
+app_regex=(
+	'Using all CPUS' \
+	'interval: 2100:3500' \
+	'deadline thread \d+' \
+	'thread\[\d+\]\s*runtime=\d+us\s*deadline=\d+us' \
+	$final_table
+	)
```
Comment (farshidtz) [inline test/rt-cyclicdeadline:28]: For the record, these are very fragile. The application's output or its output style may change from one version to the next, making it difficult to maintain tests.

**Review by farshidtz — COMMENTED** [review]
I've added some comments. Some are just for information and not requesting changes.

File: test/rt-deadline_test [hunk 6]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+# Skipping this because of: https://github.com/canonical/rt-tests-snap/issues/8
+exit 0
```
[the remark this answers is not in the document]
Comment (locnnil) [reply test/rt-deadline_test]: Right! Done!

File: test/rt-cyclicdeadline [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+. common
+
+name=$(basename "$0")
+test_init
+create_log_file
+
+printc "YELLOW" "[INFO] Standard error being directed to: $stderr_log"
+
+output=$(cyclicdeadline --quiet --interval=3500 --duration=1s --threads=$(nproc) 2>$stderr_log)
```
[the remark this answers is not in the document]
Comment (locnnil) [reply test/rt-cyclicdeadline]: Done! Thank you!

File: test/common [hunk 8]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+declare -A COLOR
+COLOR['END']='\033[0m'
+COLOR['RED']='\033[0;31m'
+COLOR['GREEN']='\033[0;32m'
+COLOR['YELLOW']='\033[0;33m'
+COLOR['BLUE']='\033[0;34m'
+COLOR['WHITE']='\033[0;37m'
+COLOR['NC']='\033[0m'
```
[the remark this answers is not in the document]
Comment (locnnil) [reply test/common]: I've changed to stronger colors and tested in lighter background and still visible.

File: test/README.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# RT-tests smoke tests 
+
+These are smoke tests that check if the main functionalities of each program is working.
+
+These tests are packaging agnostic, so doesn't matter how did you installed the `rt-tests` package.
```
Comment (farshidtz) [inline test/README.md:5]:
```suggestion
```
Actually they are packaging specific. For them to work, the snap needs aliases, and there may be other differences in other packaging forms which hasn't been tested.

File: test/common [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+# Catch the stout and stderr to different variables
+#
+catch() {
+    {
+        IFS=$'\n' read -r -d '' APP_STDOUT;
+        IFS=$'\n' read -r -d '' APP_STDERR;
+        (IFS=$'\n' read -r -d '' _ERRNO_; return ${_ERRNO_});
+    } < <((printf '\0%s\0%d\0' "$(((({ "${@}"; echo "${?}" 1>&3-; } | tr -d '\0' 1>&4-) 4>&2- 2>&1- | tr -d '\0' 1>&4-) 3>&1- | exit "$(cat)") 4>&1-)" "${?}" 1>&2) 2>&1)
```
Comment (farshidtz) [inline test/common:82]: Please add a comment explaining what this does as it isn't obvious.

File: test/common [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	[$1 -ne 0] && nonzero_fail
+	partial_success
+}
+
+test_sucess() {
+	printc "GREEN" "PASS"
+	exit 0
+}
+
+# Catch the stout and stderr to different variables
```
Comment (farshidtz) [inline test/common:75]:
```suggestion
# Catch the stdout and stderr to different variables
```

File: test/common [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+	printc "YELLOW" ${APP_STDERR}
+	printc "NC" "[INFO] Command executed successfully"
+}
+
+app_exit_check(){
+	[$1 -ne 0] && nonzero_fail
+	partial_success
+}
+
+test_sucess() {
```
Comment (farshidtz) [inline test/common:70]:
```suggestion
test_success() {
```

File: test/common [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+COLOR['END']='\033[0m'
+COLOR['RED']='\033[0;91m'
+COLOR['GREEN']='\033[0;92m'
+COLOR['YELLOW']='\033[0;93m'
+COLOR['BLUE']='\033[0;94m'
+COLOR['WHITE']='\033[0;37m'
+COLOR['NC']='\033[0m'
+COLOR['CYAN']='\033[0;96m'
+
+readonly SKIP=200
```
Comment (farshidtz) [inline test/common:13]: Hard to know what this is.
```suggestion
readonly SKIP_EXIT_CODE=200
```

File: test/common (hunk 11 above)
Comment (farshidtz) [reply test/common:75]: Doesn't this function also execute the program?

File: test/rt-deadline_test [hunk 14]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+. common
+
+# Skipping this because of: https://github.com/canonical/rt-tests-snap/issues/8
+exit $SKIP
```
Comment (farshidtz) [inline test/rt-deadline_test:6]: Would be more consistent to do this also via a function, avoiding the need for the global var and name confusion.

File: test/README.md [hunk 15]
```diff
@@ -0,0 +1,14 @@
+# RT-tests smoke tests 
+
+These are smoke tests that check if the main functionalities of each program is working.
```
Comment (farshidtz) [inline test/README.md:3]: Please also explain how.

File: test/README.md [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+
+These tests are packaging agnostic, so doesn't matter how did you installed the `rt-tests` package.
+
+
+## Running tests
+
```
Comment (farshidtz) [inline test/README.md:9]: Please explain and provide the commands for the required snap aliasing (or add commands to top-level readme and link to it).

**Review by farshidtz — CHANGES_REQUESTED** [review]
Please see my second round of review.

File: test/README.md (hunk 9 above)
Comment (locnnil) [reply test/README.md:5]: The requirement of having alias is snap specific, but not the tests itself. I can complement this talking that for the snap it is necessary to have alias.

File: test/common (hunk 11 above)
Comment (locnnil) [reply test/common:75]: @farshidtz yes! It's being executed in `"${@}";` part.

Comment (farshidtz) [reply test/common:75]: So, the comment should be adjusted, explaining this.

File: test/README.md (hunk 9 above)
Comment (farshidtz) [reply test/README.md:5]: The test suite only works if rt-tests are installed as a deb. For the snap, aliases are needed. When built from source something else needs to be done (adding them to path). And possibly different for rpm, flatpack, etc that you haven't verified.

Comment (locnnil) [reply test/README.md]: When building from source and installed using the upstream `Makefile` it's installed on `/usr/local/bin` which already is in the PATH. But I understand what are you saying!

**Review by locnnil — COMMENTED** [review]
Requested changes addressed

**Review by farshidtz — APPROVED** [review]
Looks good. Thank you.

Please avoid rewriting commits that have been reviewed. This makes it very hard to track and re-review the changes.

**Pull request merged**